torch/csrc/distributed/spmd/event_handler_impl.cpp

```diff
@@ -0,0 +1,209 @@
+          event->schema().type_);
+  }
+}
+
+std::vector<std::shared_ptr<Future>> DefaultTrigger::handlePrepareModule(
+    c10::intrusive_ptr<PrepareModuleEvent> event) {
+  params_ = event->parameters();
+  for (size_t index = 0; index < params_.size(); ++index) {
+    auto& param = params_[index];
+    auto gradAccumulator = torch::autograd::impl::grad_accumulator(param);
+    // Hook to execute after the gradient accumulator has executed.
+    gradAccumulator->add_post_hook(
+        torch::make_unique<torch::autograd::utils::LambdaPostHook>(
+            [this, index](
+                const torch::autograd::variable_list& outputs,
+                const torch::autograd::variable_list& /* unused */) {
+              autogradHook(index);
+              return outputs;
+            }));
+    gradAccumulators_.push_back(std::move(gradAccumulator));
+  }
+  return {};
+}
+
+std::vector<std::shared_ptr<Future>> DefaultTrigger::handlePreForward(
+    c10::intrusive_ptr<PreForwardEvent> /* unused */) {
+  gradReadyFutures_.clear();
+  gradReadyFutures_.reserve(params_.size());
+  for (size_t i = 0; i < params_.size(); ++i) {
+    gradReadyFutures_.emplace_back(
+        std::make_shared<Future>(at::AnyClassType::get()));
+  }
+  return gradReadyFutures_;
+}
+
+void DefaultTrigger::autogradHook(size_t index) {
+  TORCH_INTERNAL_ASSERT(gradReadyFutures_.size() == params_.size());
+  auto lgr = c10::make_intrusive<LocalGradReadyEvent>(
+      index, params_[index].mutable_grad());
+
+  gradReadyFutures_[index]->markCompleted(
+      IValue(c10::static_intrusive_pointer_cast<Event>(lgr)));
+}
+
+/////////////////////////////////////////////////////////////////////
+//                        DefaultBucketer                          //
+/////////////////////////////////////////////////////////////////////
+
+// FIXME: we might need more advanced ingress/egress event specifications.
+// E.g., LOCAL_GRAD_READY -> BUCKET_READY; COMM_DONE -> GLOBAL_GRAD_READY,
+// otherwise, DefaultBucketer and AllReduceComm can form a cycle.
+std::vector<EventSchema> DefaultBucketer::ingressEvents() const {
+  // FIXME: consume PREPARE_MODULE to allocate buckets
+  return {
+      EventType::PREPARE_MODULE,
+      EventType::LOCAL_GRAD_READY,
+      EventType::COMM_DONE};
+}
+
+std::vector<EventSchema> DefaultBucketer::egressEvents() const {
+  return {EventType::BUCKET_READY, EventType::GLOBAL_GRAD_READY};
+}
+
+std::vector<std::shared_ptr<Future>> DefaultBucketer::handleEvent(
+    const c10::intrusive_ptr<Event>& event) {
+  switch (event->schema().type_) {
+    case EventType::PREPARE_MODULE: {
+      return handlePrepareModule(
+          c10::static_intrusive_pointer_cast<PrepareModuleEvent>(event));
+    }
+    case EventType::LOCAL_GRAD_READY: {
+      return handleLocalGradReady(
+          c10::static_intrusive_pointer_cast<LocalGradReadyEvent>(event));
+    }
+    case EventType::COMM_DONE: {
+      return {};
```

Comment:
We are not handling this event?

Reply:
For the current implementation, `DefaultBucketer` directly uses `param.grad` as the bucket, so nothing else needs to be done after `COMM_DONE`. In followup PRs, when we enhance this `DefaultBucketer` to produce real buckets, it will then need to handle `COMM_DONE` to copy grads from buckets back to `param.grad` field.